### PR TITLE
Add migrate-components script

### PR DESCRIPTION
**Pull request by 14rcole**
Add script to migrate component spec.containerImage field to status.lastPromotedImage field

### CODE REVIEW

File: migrate-components.sh [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+#  - jq
+
+younger() {
+    RESOURCE_A=$1
+    RESOURCE_B=$2
+
+    # compare metadata.creationTimestamp and return whichever is younger
+}
+
+# takews two snapshot jsons and returns whichever has the newest metadata.creationTimestamp
```
Comment (jsztuka) [inline migrate-components.sh]: takews > takes

File: migrate-components.sh [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+    SNAPSHOTS=$(kubectl get snapshots -n $NAMESPACE --sort-by .metadata.creationTimestamp --selector=pac.test.appstudio.openshift.io/type=override,appstudio.openshift.io/application=$APPLICATION --no-headers | tac | cut -d ' ' -f1)
+
+    if ! [[ $SNAPSHOTS ]]; then
+        return 0
+    fi
+
+    while read -r snapshot; do
+        snapshot_json=$(kubectl get -o json snapshot $snapshot -n $NAMESPACE)
+        echo "all snapshots: $SNAPSHOTS"
```
Comment (jsztuka) [inline migrate-components.sh]: Do we still wanna list all snapshots? Even though the condition will filter only override ones? Just concerned about the log scale.

Comment (14rcole) [reply migrate-components.sh]: That and the line above were debugging output, I removed them.  Just testing the change locally now

File: migrate-components.sh [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+#
+# If spec.containerImage has been updated by imageRepository (has not sha)
+# then the script will attempt to get the latest push snapshot or override
+# snapshot that contains the component and gets the image from there
+#
+# REQUIREMENTS:
+#  - kubectl >= 1.24
+#  - jq
+
+younger() {
```
Comment (dirgim) [inline migrate-components.sh]: nit: this function seems to be unused.

File: migrate-components.sh [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+# takes two snapshot jsons and returns whichever has the newest metadata.creationTimestamp
+get_newest_snapshot_json() {
+    PUSH_SNAPSHOT_JSON=$1
+    OVERRIDE_SNAPSHOT_JSON=$2
+
+    # otherwise, return whichever is younger
+    PS_TIMESTAMP=$(echo $PUSH_SNAPSHOT_JSON | jq .metadata.creationTimestamp)
+    OS_TIMESTAMP=$(echo $OVERRIDE_SNAPSHOT_JSON | jq .metadata.creationTimestamp)
+    if [[ $OS_TIMESTAMP > $PS_TIMESTAMP ]]; then
+        echo $OVERRIDE_SNAPSHOT
```
Comment (kasemAlem) [inline migrate-components.sh]:
```suggestion
        echo "$OVERRIDE_SNAPSHOT_JSON"
```

File: migrate-components.sh [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+# takes two snapshot jsons and returns whichever has the newest metadata.creationTimestamp
+get_newest_snapshot_json() {
+    PUSH_SNAPSHOT_JSON=$1
+    OVERRIDE_SNAPSHOT_JSON=$2
+
+    # otherwise, return whichever is younger
+    PS_TIMESTAMP=$(echo $PUSH_SNAPSHOT_JSON | jq .metadata.creationTimestamp)
+    OS_TIMESTAMP=$(echo $OVERRIDE_SNAPSHOT_JSON | jq .metadata.creationTimestamp)
+    if [[ $OS_TIMESTAMP > $PS_TIMESTAMP ]]; then
```
Comment (kasemAlem) [inline migrate-components.sh:22]: To ensure a string comparison 
```suggestion
    if [[ "$OS_TIMESTAMP" > "$PS_TIMESTAMP" ]]; then
```

File: migrate-components.sh [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+        echo $OVERRIDE_SNAPSHOT
+    fi
+    echo $PUSH_SNAPSHOT
```
Comment (kasemAlem) [inline migrate-components.sh]:
```suggestion
    echo "$PUSH_SNAPSHOT_JSON"
```

File: migrate-components.sh (hunk 4 above)
Comment (14rcole) [reply migrate-components.sh]: Good catch, thanks

File: migrate-components.sh (hunk 5 above)
Comment (14rcole) [reply migrate-components.sh:22]: jq already outputs those values as strings here, I  tested this pretty thoroughly and it works as expected, so I don't want to try quoting the variables in cases that changes the behavior (it shouldn't, but I don't really trust bash)

**Review by dirgim — APPROVED** [review]
LGTM

**Review by kasemAlem — APPROVED** [review]

**Pull request merged**